viewer/packages/cad-parsers/src/utilities/SectorScene.ts

```diff
@@ -88,60 +88,114 @@ export class SectorSceneImpl implements SectorScene {
         }
-        return max;
-      },
-      { count: 0, idx: -1 }
-    ).idx;
-    clusters.idxs.forEach((cluster, idx) => {
-      clusterCounts[cluster]++;
-      clusterBounds[cluster].expandByPoint(allBounds[idx].min);
-      clusterBounds[cluster].expandByPoint(allBounds[idx].max);
-    });
-
-    const intersectingBounds = clusterBounds.filter((x, idx) => {
-      if (idx !== biggestCluster && x.intersectsBox(clusterBounds[biggestCluster])) {
         return true;
+      });
+
+      // If more than 5% of the geometry is located in the root we return the full BoundingBox.
+      const rootGeomeryThreshold = allBoundsRenderCost * 0.05;
+      if (this.root.estimatedRenderCost > rootGeomeryThreshold) {
+        return this.root.subtreeBoundingBox;
       }
-      return false;
+    }
+
+    //Sort by diagonal length, ascending
+    allBounds.sort((a, b) => {
+      return a.min.distanceTo(a.max) - b.min.distanceTo(b.max);
     });
-    if (intersectingBounds.length > 0) {
-      // Overlapping clusters - assume it's because the model doesn't contain junk geometry
-      const merged = clusterBounds[biggestCluster].clone();
-      intersectingBounds.forEach(x => {
-        merged.expandByPoint(x.min);
-        merged.expandByPoint(x.max);
+
+    //Discard 1% of largest bounds
+    const validBounds = allBounds.slice(0, Math.ceil(allBounds.length * 0.99));
+
+    //Find valid corners and create a merged bonds consisting of all valid bounds
+    const corners: number[][] = [];
+    {
+      const validBoundsMerged = new THREE.Box3();
+      validBounds.forEach(x => {
+        corners.push(x.min.toArray(), x.max.toArray());
+        validBoundsMerged.union(x);
+      });
+
+      // Create a centroid for each "corner" of the validBounds + one for the center.
+      // This is mostly done because skmeans would return different clusters when it
+      // was allowed to pick centroids itself.
+      const { min, max } = validBoundsMerged;
+      const centerZ = min.z + (max.z - min.z) * 0.5;
+      const centroids: number[][] = [
+        [min.x, min.y, centerZ],
+        [min.x, max.y, centerZ],
+        [max.x, min.y, centerZ],
+        [max.x, max.y, centerZ],
+        [min.x + (max.x - min.x) * 0.5, min.y + (max.y - min.y) * 0.5, centerZ]
+      ];
+
+      // Cluster the corners into groups and determine bounds of each cluster
+      const numClusters = Math.min(corners.length, centroids.length);
+      const clusters = skmeans(corners, numClusters, centroids, 10);
+      const clusterCounts = new Array<number>(numClusters).fill(0);
+      const clusterBounds = clusterCounts.map(_ => new THREE.Box3());
+      clusters.idxs.map(x => clusterCounts[x]++);
+      const biggestCluster = clusterCounts.reduce(
+        (max, count, idx) => {
+          if (count > max.count) {
+            max.count = count;
+            max.idx = idx;
+          }
+          return max;
+        },
+        { count: 0, idx: -1 }
+      ).idx;
```

Comment:
```suggestion
const biggestCluster = clusterCounts.reduce(
    (max, count, idx) => count > max.count ? { count, idx } : max, 
{ count: 0, idx: -1 }).idx
```

Reply:
It will not look this nice once linting is done with it, but still prettier than before 👍